Index: linux-3.10.14/drivers/mtd/nand/nand_device_list.h
===================================================================
--- linux-3.10.14.orig/drivers/mtd/nand/nand_device_list.h	2020-08-25 17:33:45.825953596 +0800
+++ linux-3.10.14/drivers/mtd/nand/nand_device_list.h	2020-08-31 10:23:51.239804551 +0800
@@ -7,6 +7,7 @@ static const flashdev_info gen_FlashTabl
     {0x92F1, 0x8095FF, 4, 8, 128, 128, 2048, 64, 0x30C77fff, "F59L1G81A", 0},
 	{0xC8DA, 0x909544, 5, 8, 256, 128, 2048, 64, 0x30C77fff, "F59L2G81A", 0},
 	{0xC8DC, 0x909554, 5, 8, 512, 128, 2048, 64, 0x30C77fff, "F59L4G81A", 0},
+	{0xEFF1, 0x009500, 4, 8, 128, 128, 2048, 64, 0x30C77fff, "W29N01HV", 0},
 	{0xECD3, 0x519558, 5, 8, 1024, 128, 2048, 64, 0x44333, "K9K8G8000", 0},
 	{0xC2F1, 0x801DC2, 4, 8, 128, 128, 2048, 64, 0x30C77fff, "MX30LF1G08AA", 0},
 	{0xC2F1, 0x809502, 4, 8, 128, 128, 2048, 64, 0x30C77fff, "MX30LF1G18AC", 0},		//LTE7240 NAND Flash.
