--- a/fsck/fsck.c
+++ b/fsck/fsck.c
@@ -45,6 +45,10 @@
 #include <dirent.h>
 #include <blkid.h>
 
+#ifdef CONFIG_TARGET_quectel
+#include <sys/sysmacros.h>
+#endif
+
 #include "fsprobe.h"
 
 #include "nls.h"
--- a/lib/sysfs.c
+++ b/lib/sysfs.c
@@ -2,6 +2,10 @@
  * Copyright (C) 2011 Karel Zak <kzak@redhat.com>
  */
 
+#ifdef CONFIG_TARGET_quectel
+#include <sys/sysmacros.h>
+#endif
+
 #include "c.h"
 #include "at.h"
 #include "pathnames.h"
--- a/libblkid/src/devname.c
+++ b/libblkid/src/devname.c
@@ -34,6 +34,10 @@
 #endif
 #include <time.h>
 
+#ifdef CONFIG_TARGET_quectel
+#include <sys/sysmacros.h>
+#endif
+
 #include "blkidP.h"
 
 #include "canonicalize.h"		/* $(top_srcdir)/include */
--- a/libblkid/src/devno.c
+++ b/libblkid/src/devno.c
@@ -32,6 +32,11 @@
 #include <fcntl.h>
 #include <inttypes.h>
 
+#ifdef CONFIG_TARGET_quectel
+
+#include <sys/sysmacros.h>
+#endif
+
 #include "blkidP.h"
 #include "pathnames.h"
 #include "at.h"
--- a/libblkid/src/evaluate.c
+++ b/libblkid/src/evaluate.c
@@ -22,6 +22,11 @@
 #include <stdint.h>
 #include <stdarg.h>
 
+#ifdef CONFIG_TARGET_quectel
+
+#include <sys/sysmacros.h>
+#endif
+
 #include "pathnames.h"
 #include "canonicalize.h"
 
--- a/libblkid/src/partitions/partitions.c
+++ b/libblkid/src/partitions/partitions.c
@@ -20,6 +20,11 @@
 #include <inttypes.h>
 #include <stdarg.h>
 
+#ifdef CONFIG_TARGET_quectel
+
+#include <sys/sysmacros.h>
+#endif
+
 #include "partitions.h"
 #include "sysfs.h"
 
--- a/libblkid/src/topology/dm.c
+++ b/libblkid/src/topology/dm.c
@@ -19,6 +19,11 @@
 #include <sys/types.h>
 #include <unistd.h>
 
+#ifdef CONFIG_TARGET_quectel
+
+#include <sys/sysmacros.h>
+#endif
+
 #include "topology.h"
 
 static int is_dm_device(dev_t devno)
--- a/libblkid/src/topology/evms.c
+++ b/libblkid/src/topology/evms.c
@@ -20,6 +20,11 @@
 #include <sys/types.h>
 #include <unistd.h>
 
+#ifdef CONFIG_TARGET_quectel
+
+#include <sys/sysmacros.h>
+#endif
+
 #include "topology.h"
 
 #define EVMS_MAJOR		117
--- a/libblkid/src/topology/md.c
+++ b/libblkid/src/topology/md.c
@@ -20,6 +20,11 @@
 #include <sys/types.h>
 #include <unistd.h>
 
+#ifdef CONFIG_TARGET_quectel
+
+#include <sys/sysmacros.h>
+#endif
+
 #include "topology.h"
 
 #ifndef MD_MAJOR
--- a/libmount/src/fs.c
+++ b/libmount/src/fs.c
@@ -15,6 +15,11 @@
 #include <blkid.h>
 #include <stddef.h>
 
+#ifdef CONFIG_TARGET_quectel
+
+#include <sys/sysmacros.h>
+#endif
+
 #include "mountP.h"
 #include "strutils.h"
 
--- a/libmount/src/tab_parse.c
+++ b/libmount/src/tab_parse.c
@@ -16,6 +16,11 @@
 #include <dirent.h>
 #include <fcntl.h>
 
+#ifdef CONFIG_TARGET_quectel
+
+#include <sys/sysmacros.h>
+#endif
+
 #include "at.h"
 #include "mangle.h"
 #include "mountP.h"
--- a/misc-utils/findmnt.c
+++ b/misc-utils/findmnt.c
@@ -33,6 +33,11 @@
 
 #include <libmount.h>
 
+#ifdef CONFIG_TARGET_quectel
+
+#include <sys/sysmacros.h>
+#endif
+
 #include "pathnames.h"
 #include "nls.h"
 #include "c.h"
--- a/misc-utils/lsblk.c
+++ b/misc-utils/lsblk.c
@@ -45,6 +45,11 @@
 #include <libudev.h>
 #endif
 
+#ifdef CONFIG_TARGET_quectel
+
+#include <sys/sysmacros.h>
+#endif
+
 #include <assert.h>
 
 #include "pathnames.h"
--- a/sys-utils/mountpoint.c
+++ b/sys-utils/mountpoint.c
@@ -34,6 +34,11 @@
 
 #include <libmount.h>
 
+#ifdef CONFIG_TARGET_quectel
+
+#include <sys/sysmacros.h>
+#endif
+
 #include "nls.h"
 #include "xalloc.h"
 #include "c.h"
--- a/libblkid/src/topology/lvm.c
+++ b/libblkid/src/topology/lvm.c
@@ -19,6 +19,11 @@
 #include <sys/types.h>
 #include <unistd.h>
 
+#ifdef CONFIG_TARGET_quectel
+
+#include <sys/sysmacros.h>
+#endif
+
 #include "topology.h"
 
 #ifndef LVM_BLK_MAJOR
--- a/disk-utils/raw.c
+++ b/disk-utils/raw.c
@@ -23,6 +23,10 @@
 #include <linux/major.h>
 #include "nls.h"
 
+#ifdef CONFIG_TARGET_quectel
+
+#include <sys/sysmacros.h>
+#endif
 
 #define RAWDEVDIR "/dev/raw/"
 #define RAWDEVCTL RAWDEVDIR "rawctl"
--- a/lib/loopdev.c
+++ b/lib/loopdev.c
@@ -31,6 +31,11 @@
 #include <dirent.h>
 #include <linux/posix_types.h>
 
+#ifdef CONFIG_TARGET_quectel
+
+#include <sys/sysmacros.h>
+#endif
+
 #include "linux_version.h"
 #include "c.h"
 #include "sysfs.h"
--- a/login-utils/login.c
+++ b/login-utils/login.c
@@ -56,6 +56,11 @@
 #include <security/pam_misc.h>
 #include <sys/sendfile.h>
 
+#ifdef CONFIG_TARGET_quectel
+
+#include <sys/sysmacros.h>
+#endif
+
 #ifdef HAVE_LIBAUDIT
 # include <libaudit.h>
 #endif
--- a/partx/partx.c
+++ b/partx/partx.c
@@ -22,6 +22,11 @@
 
 #include <blkid.h>
 
+#ifdef CONFIG_TARGET_quectel
+
+#include <sys/sysmacros.h>
+#endif
+
 #include "c.h"
 #include "pathnames.h"
 #include "nls.h"
