diff -urN dropbear.orig/default_options.h dropbear/default_options.h
--- dropbear.orig/default_options.h	2019-04-19 10:51:52.011144204 +0800
+++ dropbear/default_options.h	2019-04-19 11:03:44.287152437 +0800
@@ -179,7 +179,11 @@
 
 /* Authentication Types - at least one required.
    RFC Draft requires pubkey auth, and recommends password */
+#ifdef TELIA_CUSTOMIZATION
+#define DROPBEAR_SVR_PASSWORD_AUTH 0
+#else
 #define DROPBEAR_SVR_PASSWORD_AUTH 1
+#endif
 
 /* Note: PAM auth is quite simple and only works for PAM modules which just do
  * a simple "Login: " "Password: " (you can edit the strings in svr-authpam.c).
